### PR TITLE
Dynamically load libfm-qt on demand to create the file dialog helper.

### PR DESCRIPTION
**Pull request by PCMan**
* This prevent the hard build-time and run-time dependency on libfm-qt.
* This speed up the loading of the QPA plugin and also avoid loading libfm-qt in Qt programs having QT_NO_GLIB=1.

This requires: https://github.com/lxqt/libfm-qt/pull/243

### CODE REVIEW

**Comment by PCMan** [comment]
@agaida @tsimonq2  With this change, building lxqt-qtplugin no longer require libfm-qt.
So people who don't like libfm-qt can live safely without it.
Qt programs that don't need the file dialogs can avoid loading libfm-qt as well.

However, I'd really suggest that you add libfm-qt to the "recommended" for your lxqt-qtplugin *.deb packages to make them installed together by default since with libfm-qt you can have a much more functional file dialog.

**Comment by PCMan** [comment]
@agaida  @tsujan @yan12125 As mentioned earlier, with this PR all future ABI changes in libfm-qt won't require the rebuild of lxqt-qtplugin. :-)

File: src/lxqtplatformtheme.cpp [hunk 1]
```diff
@@ -222,8 +228,32 @@ bool LXQtPlatformTheme::usePlatformNativeDialog(DialogType type) const {
...
     if(type == FileDialog
        && qobject_cast<QApplication *>(QCoreApplication::instance())) { // QML may not have qApp
-        // use our own file dialog
-        return new LXQtFileDialogHelper();
+        // use our own file dialog provided by libfm
+
+        // When a process has this environment set, that means glib event loop integration is disabled.
+        // In this case, libfm-qt just won't work. So let's disable the file dialog helper and return nullptr.
+        if(qgetenv("QT_NO_GLIB") == "1") {
+            return nullptr;
```
Comment (yan12125) [inline src/lxqtplatformtheme.cpp]: How about adding a message to let users know why the improved file dialog is unavailable?

Comment (PCMan) [reply src/lxqtplatformtheme.cpp]: We can add a console log for it but I'm not sure if this is needed.

**Comment by agaida** [comment]
@PCMan - there is a problem with Recommend: - some really clever guy do 
```
apt install $foo --no-install-recommends
```
and are whining about missed functionality afterwards - as long they stay with pcmanfm-qt - no problem. And i really don't care about Installations without our main components, so a Recommend: libfm-qt? is the way to go.

File: src/lxqtplatformtheme.cpp (hunk 1 above)
Comment (agaida) [reply src/lxqtplatformtheme.cpp]: console output will go unreaded to .xsession-errors - please don't

**Comment by tsujan** [comment]
@agaida I think recommending libfm-qt is enough. The "worst" thing that could happen is that the user won't see a feature-rich file dialog without libfm-qt.

I'll test the changes tomorrow.

**Comment by agaida** [comment]
@tsujan - recommend is enough - and as long the users don't build the system from the scratch it would be damn hard to avoid libfm-qt and pcmanfm-qt firsthand.

**Comment by agaida** [comment]
@PCMan - Re: needed rebuilds - never did it by hand for debian - thats what transitions are for. But you are right, that should be mentioned in the packaging notes with a big fat exclamation mark. It is important for all our friends who build directly from git (or use the AUR or similar things).

**Comment by luis-pereira** [comment]
Will review and test Monday, only.

--
Cumprimentos
Regards

Em 28/07/2018 22:45, "Alf Gaida" <notifications@github.com> escreveu:

@PCMan <https://github.com/PCMan> - Re: needed rebuilds - never did it by
hand for debian - thats what transitions are for. But you are right, that
should be mentioned in the packaging notes with a big fat exclamation mark.
It is important for all our friends who build directly from git (or use the
AUR or similar things).

—
You are receiving this because your review was requested.

Reply to this email directly, view it on GitHub
<https://github.com/lxqt/lxqt-qtplugin/pull/39#issuecomment-408636671>, or mute
the thread

**Comment by tsujan** [comment]
@agaida I know this is off topic and so, I won't continue it here but, IMO, an LXQt meta package that provides a good LXQt experience could solve your problem with those "clever guys".

**Comment by agaida** [comment]
@tsujan: not exactly - as long the clever guys use at least lxqt-core without recommends - all things are fine. If this package is still to much "bloat" and they use the packages without the meta-packages things might go south. But thats not a common support case - in case they use the packages alone they should know what they are doing - at least i hope so. So no problem at all.

**Comment by tsujan** [comment]
Applied both patches. Unfortunately, all Qt apps show the the default Qt file dialog; there's no trace of LXQt file dialog anymore.

**Comment by tsujan** [comment]
@agaida Do you see the same thing?

**Comment by tsujan** [comment]
@agaida If you mean you get it with all Qt apps, I should have made a mistake somewhere...

**Comment by agaida** [comment]
works with nomacs, lximage-qt, vokoscreen, virtualbox but not not with featherpad

**Comment by agaida** [comment]
ah - featherpad has't had the hook in systemdialog

**Comment by tsujan** [comment]
@agaida It's lxqt-archiver in your screenshot; it used lxqt file dialog explicitly. 

I hadn't make a mistake; something's definitely wrong.

**Comment by tsujan** [comment]
Falkon (formerly QupZilla), FeatherPad and FeatherNotes use the Qt file dialog, while they use the native file dialog under KDE and Gnome -- they also use the LXQt file dialog after I reversed the changes a few minutes ago.

**Comment by PCMan** [comment]
@tsujan do you install libraries to custom dirs? It's possible that libfmqt is not found by qlibrary.

You can use LD_DEBUG=all env var to see what's happening.

What I did is just dynamically load the lib at runtime.

**Comment by tsujan** [comment]
> do you install libraries to custom dirs?

No I don't. @agaida sees the same issue with some Qt apps.

**Comment by agaida** [comment]
@tsujan - no, but fp was set to use the native dialog

**Comment by agaida** [comment]
So - for me there is only one remaining problem that has nothing to do with the current PRs. Most of the file open dialogs are full screen and not sizable - guess i had filed a bug about before.

**Comment by agaida** [comment]
and i still have a problem with linking fm_qt in src/CMakeLists.txt - one  will not notice that with local builds, but will do so in a build chroot.

**Comment by tsujan** [comment]
> no, but fp was set to use the native dialog

You mean it was set to NOT use the native dialog -- which isn't the default.

I tried with every Qt app I had; the theory was proved in the practice: the library was loaded and the symbol was resolved.

**Comment by tsujan** [comment]
> Most of the file open dialogs are full screen and not sizable 

That should be related to the WM you use: it makes large windows full-screen. Close all file dialogs, remove `~/.config/lxqt/filedialog.conf` (or just change `WindowSize` inside it) and retry.

If my guess is correct, we might need to prevent LXQt file dialog from becoming full-screen OR add F11 to it for making it (un-)full-screen.

**Comment by agaida** [comment]
@tsujan - you was right about the dialog, deleted the size line, now it work as expected.
right now is `WindowSize=@Size(1041 738)`  - and the Window is sizable again - and i found a very special kind of bug - will make some screenshots, hard to explain.

**Comment by tsujan** [comment]
@agaida Let's discuss it at https://github.com/lxqt/libfm-qt/issues/244

**Comment by PCMan** [comment]
@agaida Thanks for the reminder.
I forgot to delete the explicit "fm-qt linkage" in src/CMakeLists.txt. Just fixed it.
It should build without problems now.

**Comment by yan12125** [comment]
Sorry for the late participation! I have no access to my Linux box during the weekend. After updating both libfm-qt and lxqt-qtplugin and tested Quassel and KeepassXC, the file choosing dialog looks as awesome as always. 

Now here's the question regarding packaging. I'm maintaining the AUR package [lxqt-qtplugin-git](https://aur.archlinux.org/packages/lxqt-qtplugin-git) for Arch Linux users. Currently `libfm-qt-git` is a hard dependency. Do you think it's a good idea to move that to a optional dependency? Note that by default, `pacman` (the package manager of Arch Linux) does not install optional dependencies.

**Comment by tsujan** [comment]
@yan12125 Yes, libfm-qt should be an optional dependency now. The problem of installing optional dependencies in Arch is like that of installing Recommends in Debian ;) Not our problem though.

**Comment by yan12125** [comment]
Got it and done!
https://aur.archlinux.org/cgit/aur.git/commit/?h=lxqt-qtplugin-git&id=fce9af42d44d76a9a6aa00429c309fbc864d7d59

2018-07-30 19:43 GMT+08:00 tsujan <notifications@github.com>:

**Comment by PCMan** [comment]
@yan12125 it's better to make it optional, but mention in the document that it's highly recommended to install libfm-qt. :-)

**Comment by agaida** [comment]
Nobody read documentations - but as long one install pcmanfm-qt as filemanager libfm-qt is installed anyways - same for lximage-qt. That leave only one corner case: One don't install pcmanfm-qt nor lximage-qt and can't read - i would call this combination simply bad luck. And i guess that a Qt-filedialog is the slightest problem in that special case. :dark_sunglasses:

**Comment by tsujan** [comment]
> i would call this combination simply bad luck

Or, if we put it less politely but more realistically, the user's ignorance ;)

**Comment by tsimonq2** [comment]
Belated ack, thanks everyone.

**Comment by yan12125** [comment]
I've update the description for optional dependencies. Now Arch users will see the following texts when installing lxqt-qtplugin-git from AUR:
```
Optional dependencies for lxqt-qtplugin-git
    libfm-qt-git: for enhanced file dialog (highly recommended)
```
Anyway, I can do nothing if they ignore these lines :)

**Comment by agaida** [comment]
those who can read, write and understand have a real advantage

But: Dsylexics have more fnu.

**Pull request merged**